### DISCUSSION TITLE
TLS with cerificate as a String Resource

### DISCUSSION THREAD

**Discussion by Mgalazka2323** [davidepianca98/KMQTT#59] [Q&A] [answered]
Hello,
I tried with TLSSettings and adding proper certificate as a parameter
MQTTClient(
            mqttVersion = MQTTVersion.MQTT3_1_1,
            address = a!!,
            port = port,
            userName = u,
            password = p?.toByteArray()?.asUByteArray(),
            cleanStart = true,
            clientId = "xxxxxxxxx"
            tls = TLSClientSettings(serverCertificate = crt!!)
)
the exception is generated like this:  java.io.FileNotFoundException:  ENAMETOOLONG (File name too long)
 crt is proper certificate I think that in older version it worked but now i tried with 0.4.6, 0.4.7, 0.4.8 but without success
Can you help me with this ?
best regards
 Mateusz

**Comment by davidepianca98** [comment]
Hello, can you please give me the crt variable content? You can generate a new certificate in the same way to avoid sharing the one you are using.

**Reply by Mgalazka2323** [reply]
hello,  in attachment :) 
[ca_test.txt](https://github.com/user-attachments/files/16746511/ca_test.txt)
Best regards
 Mateusz

**Reply by davidepianca98** [reply, marked as answer]
I tried replicating the problem but I couldn't. The certificate is correctly parsed. If I change the header to corrupt the data, then I get a java.io.FileNotFoundException but not for file name too long, just incorrect path syntax, but that could be a difference of the JVM we are using to run it. Please check again that your crt variable actually contains the correct certificate value OR the path to the file containing it.

This is my code:
```
public fun main() {
    val ca = """-----BEGIN CERTIFICATE-----
MIID7zCCAtegAwIBAgIUI4f8AI+Ldm0QuuVjUT8/Lkr8orowDQYJKoZIhvcNAQEL
BQAwgYUxCzAJBgNVBAYTAlBMMRAwDgYDVQQIDAdzbGFza2llMRMwEQYDVQQHDApM
b2R5Z293aWNlMRswGQYDVQQKDBJJbnByYXggc3AuIHogby4gby4xEjAQBgNVBAMM
CWlucHJheC5wbDEeMBwGCSqGSIb3DQEJARYPYml1cm9AaW5wcmF4LnBsMCAXDTI0
MDExMDExNDQ1NFoYDzIxMjMxMjE3MTE0NDU0WjCBhTELMAkGA1UEBhMCUEwxEDAO
BgNVBAgMB3NsYXNraWUxEzARBgNVBAcMCkxvZHlnb3dpY2UxGzAZBgNVBAoMEklu
cHJheCBzcC4geiBvLiBvLjESMBAGA1UEAwwJaW5wcmF4LnBsMR4wHAYJKoZIhvcN
AQkBFg9iaXVyb0BpbnByYXgucGwwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEK
AoIBAQDPCyn95ff7YrRzrj9tsPlKuM1PKuJBPUO1I6h30phx0Xa/yG2gCT312zua
cKqJgkVgLYVLMEwaezyQ9tCnWOwhwIKmlv28zO3G2IEGjvBU1+r3viWXfWared6/
oT0nW/jzeKZvoUV/qzfU7xYlZH4qwNqgV09nB43WYPP67pseiaCxDKmtdzw+FrA6
49fPaho7m0Pu4Ks02nu4eoK5DOYbloH4K13jPH3s8VVjfKQLkDmKvw4Q6hyW1Tib
HhQ+k/AcIcjtysjy27RpWR1kRAXa03yvY/7AVAw71qIWJ+7zgwSoNEEgm/Cikat+
o14V+kJCzie6zl6NQGGjKttWSuOjAgMBAAGjUzBRMB0GA1UdDgQWBBREEniSD+9v
Qgl5nxVJpCO8VJbONjAfBgNVHSMEGDAWgBREEniSD+9vQgl5nxVJpCO8VJbONjAP
BgNVHRMBAf8EBTADAQH/MA0GCSqGSIb3DQEBCwUAA4IBAQB/KrD614UzhsqPCDYw
KD2hwJgSUPgULbIbE5+waYiYx1JQweZYmqWQdZfQVtWx1jJNp92/ptRddp54xfjY
qDwJUtYd3ZNdjGoFYI5BkmId7Pq0bHQ8wLZ6Q1eLdvsrTZzAh6QvV5C+tMY85I/8
bd4Txl720oVNkhzeiHfnxctMKnvY2o7NkU2FYa62ZpGJ65JqpoDqyvHOGwnA49KF
yE/F+CbYjAP36QpCoqKZYO94pXUMT+I74XwbD8hiC/kUKOfq3nmavLuPrVMf16Uh
nUlhE8u+UpWu7qXTNbd3+8ACJ2kNVj8Zh4ly77C6h6+4rIgSezbWnUJpvKoGiUF3
Auk0
-----END CERTIFICATE-----
"""
    val client = MQTTClient(
        MQTTVersion.MQTT3_1_1,
        "localhost",
        1883,
        TLSClientSettings(serverCertificate = ca),
        30,
        onDisconnected = {
            println("Disconnected")
        }
    ) {
        println(it.payload?.toByteArray()?.decodeToString())
    }
    client.run()
}
```

**Reply by Mgalazka2323** [reply]
Hello, thank you for your time ...you're right , problem was with my certificate because \n  after saving was removed and certificate doesn't have proper structure.
Thank you one more time for your time and great job with this library !!
Best regards
 Mateusz
